### PR TITLE
GraphQL Initial Scuffolding

### PR DESCRIPTION
**Pull request by andi-huber**
New GraphQL Viewer (Incubator) based on Spring for GraphQL 1.0.0-M5+
https://spring.io/blog/2022/01/19/spring-for-graphql-1-0-0-m5-released

### CODE REVIEW

File: .run/DemoAppWicketJdo.run.xml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<component name="ProjectRunConfigurationManager">
```
Comment (andi-huber) [inline .run/DemoAppWicketJdo.run.xml:1]: I guess we need to add .run entry with gitignore

File: core/metamodel/src/main/java/org/apache/isis/core/metamodel/methods/DomainIncludeAnnotationEnforcesMetamodelContributionValidator.java [hunk 2]
```diff
@@ -127,13 +127,20 @@ public void validate(final ObjectSpecification spec) {
                     .collect(Collectors.joining("; "));
 
             //FIXME[ISIS-2774] - update message to a more generic one
+
```
Comment (andi-huber) [inline core/metamodel/src/main/java/org/apache/isis/core/metamodel/methods/DomainIncludeAnnotationEnforcesMetamodelContributionValidator.java:130]: meanwhile fixed in 'master'

File: core/metamodel/src/main/java/org/apache/isis/core/metamodel/specloader/SpecificationCacheDefault.java [hunk 3]
```diff
@@ -100,7 +102,9 @@ public void forEach(Consumer<T> onSpec, boolean shouldRunConcurrent) {
         if(shouldRunConcurrent) {
             vList.forEachParallel(onSpec);
         } else {
-            vList.forEach(onSpec);
+            vList
+                .stream().sorted(Comparator.comparing(HasLogicalType::getLogicalTypeName))
```
Comment (andi-huber) [inline core/metamodel/src/main/java/org/apache/isis/core/metamodel/specloader/SpecificationCacheDefault.java:106]: Is there a need for sorting? Or is it a debugging thing.

File: examples/demo/wicket/jdo/src/main/java/demoapp/webapp/wicket/jdo/DemoAppWicketJdo.java [hunk 4]
```diff
@@ -78,10 +78,15 @@ public static void main(final String[] args) {
     	IsisPresets.prototyping();
         //IsisPresets.logging(WebRequestCycleForIsis.class, "debug");
 
-        System.setProperty("spring.profiles.active", "demo-jdo");
+        System.setProperty("spring.profiles.active", preservingAnyExisting("demo-jdo"));
```
Comment (andi-huber) [inline examples/demo/wicket/jdo/src/main/java/demoapp/webapp/wicket/jdo/DemoAppWicketJdo.java:81]: logic for adding a spring profile, could be refactored into a common (shared) utility;
eg. package `org.apache.isis.core.config.util`

File: incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlAutoConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+import org.springframework.graphql.execution.ExecutionGraphQlService;
+import org.springframework.graphql.execution.GraphQlSource;
+import org.springframework.graphql.execution.MissingSchemaException;
+import org.springframework.graphql.execution.RuntimeWiringConfigurer;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for creating a Spring GraphQL base
+ * infrastructure.
+ *
+ * @author Brian Clozel
```
Comment (andi-huber) [inline incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlAutoConfiguration.java:56]: authorship?

File: incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/source/ExecutionStrategyResolvingWithinInteraction.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+@Service
+@RequiredArgsConstructor(onConstructor_ = {@Inject})
+public class ExecutionStrategyResolvingWithinInteraction extends AsyncExecutionStrategy {
+
+    private final InteractionService interactionService;
+
+    @Override
+    protected CompletableFuture<FieldValueInfo> resolveFieldWithInfo(ExecutionContext executionContext, ExecutionStrategyParameters parameters) {
+
+        interactionService.openInteraction();
```
Comment (andi-huber) [inline incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/source/ExecutionStrategyResolvingWithinInteraction.java:27]: use rather ..
```
interactionService.runAnonymous(()->{
...
});
```

File: incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlCorsProperties.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+import org.springframework.boot.convert.DurationUnit;
+import org.springframework.lang.Nullable;
+import org.springframework.util.CollectionUtils;
+import org.springframework.web.cors.CorsConfiguration;
+
+/**
+ * Configuration properties for GraphQL endpoint's CORS support.
+ *
+ * @author Andy Wilkinson
+ * @author Brian Clozel
```
Comment (andi-huber) [inline incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlCorsProperties.java:35]: authorship?

File: incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlProperties.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+import java.time.Duration;
+import java.util.Arrays;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+/**
+ * {@link ConfigurationProperties properties} for Spring GraphQL.
+ *
+ * @author Brian Clozel
```
Comment (andi-huber) [inline incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlProperties.java:27]: authorship?

File: incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlSourceBuilderCustomizer.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+package org.apache.isis.viewer.graphql.viewer.spring;
+
+import org.springframework.graphql.execution.GraphQlSource;
+
+/**
+ * Callback interface that can be implemented by beans wishing to customize properties of
+ * {@link org.springframework.graphql.execution.GraphQlSource.Builder} whilst retaining
+ * default auto-configuration.
+ *
+ * @author Rossen Stoyanchev
```
Comment (andi-huber) [inline incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlSourceBuilderCustomizer.java:26]: authorship?

File: incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlWebMvcAutoConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+import org.springframework.web.socket.WebSocketHandler;
+import org.springframework.web.socket.server.support.DefaultHandshakeHandler;
+import org.springframework.web.socket.server.support.WebSocketHandlerMapping;
+import org.springframework.web.socket.server.support.WebSocketHttpRequestHandler;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for enabling Spring GraphQL over
+ * Spring MVC.
+ *
+ * @author Brian Clozel
```
Comment (andi-huber) [inline incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlWebMvcAutoConfiguration.java:72]: authorship?

File: incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/InvalidSchemaLocationsException.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+import org.springframework.core.NestedRuntimeException;
+import org.springframework.core.io.support.ResourcePatternResolver;
+import org.springframework.util.Assert;
+
+/**
+ * {@link InvalidSchemaLocationsException} thrown when no schema file could be found in
+ * the provided locations.
+ *
+ * @author Brian Clozel
```
Comment (andi-huber) [inline incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/InvalidSchemaLocationsException.java:32]: authorship?

File: incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/webmodule/WebModuleGraphql.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+ */
+@Service
+@Named("isis.viewer.graphql.WebModuleGraphql")
+@javax.annotation.Priority(PriorityPrecedence.MIDPOINT - 80)
+@Qualifier("Graphql")
+public final class WebModuleGraphql extends WebModuleAbstract {
+
+    private final IsisConfiguration isisConfiguration;
+
+    @Inject
```
Comment (andi-huber) [inline incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/webmodule/WebModuleGraphql.java:66]: Use `@Autowired` instead, as stated with the java-doc. Should one ever want to use this with the JEE stack (CDI).

File: .run/DemoAppWicketJdo.run.xml (hunk 1 above)
Comment (danhaywood) [reply .run/DemoAppWicketJdo.run.xml:1]: Actually, this directory holds an IntelliJ run configuration, so is useful for those of us using that IDE.

File: core/metamodel/src/main/java/org/apache/isis/core/metamodel/specloader/SpecificationCacheDefault.java (hunk 3 above)
Comment (danhaywood) [reply core/metamodel/src/main/java/org/apache/isis/core/metamodel/specloader/SpecificationCacheDefault.java:106]: It's a presentation thing in the graphQL metamodel - seems reasonable behaviour?

File: examples/demo/wicket/jdo/src/main/java/demoapp/webapp/wicket/jdo/DemoAppWicketJdo.java (hunk 4 above)
Comment (danhaywood) [reply examples/demo/wicket/jdo/src/main/java/demoapp/webapp/wicket/jdo/DemoAppWicketJdo.java]: rebased on master and picked up improved handling of this.

File: incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/source/ExecutionStrategyResolvingWithinInteraction.java (hunk 6 above)
Comment (danhaywood) [reply incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/source/ExecutionStrategyResolvingWithinInteraction.java:32]: Done.

File: incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlProperties.java (hunk 8 above)
Comment (danhaywood) [reply incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlProperties.java:27]: We've nicked some code from Spring Boot 2.7 milestone, so this class will be deleted in due course.  Seems reasonable to maintain authorship for now.

File: incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlSourceBuilderCustomizer.java (hunk 9 above)
Comment (danhaywood) [reply incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlSourceBuilderCustomizer.java:26]: ditto

File: incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlWebMvcAutoConfiguration.java (hunk 10 above)
Comment (danhaywood) [reply incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/GraphQlWebMvcAutoConfiguration.java:72]: ditto

File: incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/InvalidSchemaLocationsException.java (hunk 11 above)
Comment (danhaywood) [reply incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/spring/InvalidSchemaLocationsException.java:32]: ditto

File: incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/webmodule/WebModuleGraphql.java (hunk 12 above)
Comment (danhaywood) [reply incubator/viewers/graphql/viewer/src/main/java/org/apache/isis/viewer/graphql/viewer/webmodule/WebModuleGraphql.java]: (file since removed)

**Pull request merged**